external/kafka/src/test/scala/org/apache/spark/streaming/kafka/KafkaSourceSuite.scala

```diff
@@ -0,0 +1,117 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.streaming.kafka
+
+import kafka.common.TopicAndPartition
+import org.apache.kafka.clients.producer.RecordMetadata
+import org.scalatest.time.SpanSugar._
+
+import org.apache.spark.sql.StreamTest
+import org.apache.spark.sql.execution.streaming.{Offset, Source}
+import org.apache.spark.sql.test.SharedSQLContext
+
+
+class KafkaSourceSuite extends StreamTest with SharedSQLContext {
```

Comment:
The problem was `AddKafkaData` did not return correct offsets if there no data was actually added, as it used the offsets returned on every successful push. Changed that. 

And adding the more complicated stress test as well with multiple topics.